Updater/Server.cs

```diff
@@ -35,312 +41,439 @@ public void Start(string ip, string port)
+    private void SyncUp(string clientId)
     {
-        public string clientID = "";
-        private readonly ICommunicator _communicator;
-        private readonly Dictionary<string, TcpClient> clientConnections = new Dictionary<string, TcpClient>(); // Track clients
-        private static int clientCounter = 0;
+        try
+        {
+            string serializedSyncUpPacket = Utils.SerializedSyncUpPacket();
 
-        public ClientMetadataHandler(ICommunicator communicator)
+            // Write equivalent of this: 
+            // ReceiveData("Syncing Up with the server");
+            Trace.WriteLine($"[Updater] Sending SyncUp request dataPacket to client: {clientId}");
+            if (_communicator != null)
+            {
+                _communicator.Send(serializedSyncUpPacket, "ServerNotificationHandler", clientId);
+            }
+            else
+            {
+                UpdateUILogs("Communicator is null");
+            }
+        }
+        catch (Exception ex)
         {
-            _communicator = communicator;
+            UpdateUILogs($"Error in SyncUp: {ex.Message}");
         }
 
-        public static void PacketDemultiplexer(string serializedData, ICommunicator communicator, string clientID)
+    }
+
+    /// <summary>
+    /// Implementation of Wait before SyncUp
+    /// </summary>
+    public void RequestSyncUp(string clientId)
+    {
+        try
         {
-            try
-            {
-                // Deserialize data
-                DataPacket dataPacket = Utils.DeserializeObject<DataPacket>(serializedData);
+            semaphore.Wait();
+            SyncUp(clientId);
+        }
+        catch (Exception ex)
+        {
+            UpdateUILogs($"Error in RequestSyncUp: {ex.Message}");
+        }
+    }
 
-                // Check PacketType
-                switch (dataPacket.DataPacketType)
-                {
-                    case DataPacket.PacketType.Metadata:
-                        MetadataHandler(dataPacket, communicator, clientID);
-                        break;
-                    case DataPacket.PacketType.ClientFiles:
-                        ClientFilesHandler(dataPacket, communicator, clientID);
-                        break;
-                    default:
-                        throw new Exception("Invalid PacketType");
-                }
-            }
-            catch (Exception ex)
+
+    /// <summary>
+    /// Complete the sync by Signalling semaphore
+    /// </summary>
+    public void CompleteSync()
+    {
+        try
+        {
+            semaphore.Signal();
+        }
+        catch (Exception ex)
+        {
+            UpdateUILogs($"Error in CompleteSync: {ex.Message}");
+        }
+    }
+
+    public static void UpdateUILogs(string message)
+    {
+        NotificationReceived?.Invoke(message);
+    }
+}
+
+/// <summary>
+/// INotificationHandler Implementation in Server side
+/// </summary>
+public class ServerNotificationHandler : INotificationHandler
```

Comment:
The notifier name should be same on both server and client side, pls check the networking code to understand better